lib/Sema/CSApply.cpp

```diff
@@ -2514,6 +2514,65 @@ namespace {
             /*implicit=*/expr->isImplicit(), Type(), getType);
         result = finishApply(apply, Type(), cs.getConstraintLocator(expr));
       }
+			
+      // If the base type is optional and we're referring to an nominal
+      // type member via the dot syntax and the member name matches
+      // Optional<T>.none then emit a diagnostic
+      if (auto baseUnwrappedType = baseTy->lookThroughAllOptionalTypes()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(result)) {
+          
+          Decl *structOrEnumMember = nullptr;
+          StringRef memberName;
+          auto calledValue = DSCE->getCalledValue();
+          bool isOptional = false;
+          
+          if (auto EED = dyn_cast<EnumElementDecl>(calledValue)) {
+            structOrEnumMember = EED;
+            memberName = EED->getNameStr();
+            isOptional = EED->getParentEnum()->isOptionalDecl() &&
+                         !EED->getParentEnum()->getElement(EED->getName())
```

Comment:
I think `EED->getParentEnum()->getElement(EED->getName())` is identical to just `EED`.